app/mailers/mailer.rb

```diff
@@ -8,16 +8,24 @@ class Mailer < ApplicationMailer
 
   def email_reset(user)
     @user = User.find(user["id"])
-    mail to: @user.unconfirmed_email,
-         subject: I18n.t("mailer.confirmation_subject",
-           default: "Please confirm your email address with RubyGems.org")
+    if @user.confirmation_token
+      mail to: @user.unconfirmed_email,
+          subject: I18n.t("mailer.confirmation_subject",
+            default: "Please confirm your email address with RubyGems.org")
+    else
+      Rails.logger.info("[rails:mailer:email_reset] confirmation token not found. skipping sending mail for #{@user.handle}")
+    end
```

Comment:
Ideally, we should not queue duplicate jobs and we won't this `if .. else` here. delayed_job considers this an application level concern collectiveidea/delayed_job#192

[profile delete](https://github.com/rubygems/rubygems.org/blob/master/app/controllers/profiles_controller.rb#L38-L42) has same issue, as in duplicate jobs fail. The above-mentioned issue has [using handler to find duplicates](https://gist.github.com/synth/fba7baeffd083a931184) as possible solution, seems a bit hacky.

Does anyone know any cleaner solution for this?

Reply:
IMHO it would make sense to wrap email reset delivery into simple background job handling this logic intead of making mailer responsible for this.

Reply:
Thanks :yellow_heart: I have updated this to use a custom job. 
I am indecisive about one more thing.. `email_reset` and `email_confirmation` can use a single job and mailer but it will need two `if` statements (for email and template). I am not sure if merging this to one function is a good idea. pretty sure codeclimate will complain about duplicate code with the current state.